### PR TITLE
docs: add a glossary

### PR DESCRIPTION
**Pull request by martinvonz**
The need for a glossary came up on Discord today.

This needs some more work, but I think this is a good start. I'm still happy to update it if anyone has suggestions, of course. I haven't started sprinkling pointers to the glossary from other places, so users will only discover it by browsing the `docs/` directory.

# Checklist

If applicable:
- [ ] I have updated `CHANGELOG.md`
- [ ] I have updated the documentation (README.md, docs/, demos/)
- [ ] I have updated the config schema (src/config-schema.json)
- [ ] I have added tests to cover my changes

### CODE REVIEW

File: docs/glossary.md
Comment (necauqua) [inline docs/glossary.md:29]:
```suggestion
A change ID is a unique identifier for a [change](#change). They are typically
```

Comment (martinvonz) [reply docs/glossary.md]: Fixed, thanks.

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:5]: "named branches" instead of "branches"

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:11]: A backend is an implementation of the storage layer. Currently, the setup jj usually uses is to use a Git repository to store commit contents (this is the "commit backend") and a custom backend to store [the operation log](#operation-log). Another bare-bones commit backend exists for testing purposes only.

Alternative backends could be used, for example, if somebody wanted to use `jj` with a humongous monorepo.

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:16]: I would call this "Named branch" and start the description with `A named branch, often referred to simply as a "branch"...`

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:30]: I'd add: `... 16 bytes long and are often randomly generated`

And I'd replace the next sentences:

> By default, `jj log` presents them as a sequence of 12 letters in the `k-z` range, at the beginning of a line. These are actually hexadecimal numbers that use "digits" `k-z` instead of `0-9a-f`.

**Review by ilyagr — COMMENTED** [review]
Looks great! I had many ideas for rewriting paragraphs, but they are all minor suggestions. I reviewed this up to the letter "C" so far. (There's no need to wait for me to get through the whole thing if somebody else is faster).

File: docs/glossary.md
Comment (listx) [inline docs/glossary.md:7]: The language "such commits" is a little surprising because commits are not discussed yet in this paragraph.

**Review by listx — COMMENTED** [review]
This looks great! Maybe also explain "root commit" (and how it has commit ID 00000)?

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Done

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Good point, rephrased.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Thanks. I copied some parts and rewrote other parts. I feel like "custom backend" is equivalent to "out-of-tree backend", which is not what we mean here.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:21]: I think we almost always call them just "branches". I added a note about "named branches" instead. Let me know what you think.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Done

**Review by martinvonz — COMMENTED** [review]
Thanks for reviewing

**Comment by martinvonz** [comment]
> Maybe also explain "root commit" (and how it has commit ID 00000)?

Good point. I added an entry for that.

File: docs/glossary.md
Comment (chooglen) [inline docs/glossary.md:5]: > An anonymous branch is a branch

I think we still need to define what the second "branch" means in this context. Since we mean "branch" in a topological sense, maybe:

```
An anonymous branch is a chain of commits that does not have any named branches pointing to it or to any of its descendants...
```

File: docs/glossary.md
Comment (chooglen) [inline docs/glossary.md:103]: Hm, does this match how most people use the word? I think people typically refer to everything on disk (e.g. the storage backend, working copy) as the "repository".

e.g. "My repository is corrupted, better delete it".

File: docs/glossary.md
Comment (chooglen) [inline docs/glossary.md:122]: "change" is a bit misleading, since we're creating a new commit and not replacing the current commit. This is addressed later, but maybe we can head off confusion with a small rewording, maybe:

```
To "rewrite" a commit means to create a new version of that commit with different contents, metadata, or both. Rewriting a commit results in a new commit...
```

File: docs/glossary.md
Comment (chooglen) [inline docs/glossary.md:145]: I think "head"/"visible head" needs definition.

**Review by chooglen — COMMENTED** [review]
I've tried not to nitpick wording or ask for extra details, since we can always enhance that in follow up PRs. Instead, I tried to focus on places where the definition seems misleading or incomplete.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Oops, I meant to write something like that but clearly forgot. Thanks.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Good point, done.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: I've tried to clarify it under #View. I actually had a "Visible heads" as its own section before but I just ended up with a circular definition and didn't know how to explain it better. Let me know if you have any suggestions.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:104]: I don't think of the working copy as part of the repository, but you may be right that most people do. I've rephrased it as "everything under `.jj/`", so it's at least not the same as the definition of "workspace".

File: docs/glossary.md
Comment (chooglen) [inline docs/glossary.md:87]: Suggestion: add a link to #visible-commit

File: docs/glossary.md
Comment (chooglen) [reply docs/glossary.md:104]: Sounds good. Have you pushed that change?

File: docs/glossary.md
Comment (chooglen) [inline docs/glossary.md:154]: Suggestion: maybe it would be useful to use the analogy "Commit is to Tree as Operation is to View".

File: docs/glossary.md
Comment (chooglen) [inline docs/glossary.md:150]: For now, I _think_ it's enough to define "head" and leave "visible" for another time (I think the latter would need a more in-depth explanation).

For "head", maybe:

```
A head is a commit with no descendants. The [view](#view) records which anonymous heads
(heads without a branch pointing to them) are visible at a given [operation](#operation).
```

File: docs/glossary.md
Comment (chooglen) [inline docs/glossary.md:146]: Suggestion: maybe visible is something like "can be returned by a revset"? I guess that's not entirely true, since you can return non-visible commits by querying by the commit id.

Maybe visible is closer to "the set of commits reachable from the current view". Then to make it non-circular, we'd drop the word "visible" from the view section, e.g.

> A view is a snapshot of branches and their targets, ~visible~ anonymous heads, and working-copy commits. The ~visible~ heads define which commits are [visible](#visible-commits).

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Now I have :)

File: docs/glossary.md
Comment (joyously) [reply docs/glossary.md:154]: >  "Commit is to Tree as Operation is to View"

The object is to explain things, not take an IQ test. This makes no sense to me.

File: docs/glossary.md
Comment (listx) [inline docs/glossary.md:79]: Remove?

File: docs/glossary.md
Comment (listx) [inline docs/glossary.md:119]: Do all revsets resolve to sets of revisions? If so maybe a better way to explain it would be "Revset stands for revision set" and then also add the functional language details. If the functional language itself has a name, add that also.

It may be too late now, but I find it a little bit confusing to use the word "revset" to mean 2 things --- both the expressions in the functional language used to select sets of revisions, but also to mean "a set of revisions" that result from evaluating the expression(s) of this same functional language.

File: docs/glossary.md
Comment (listx) [inline docs/glossary.md:134]: The "where root commits are ..." part is tricky to understand --- it's not immediately clear if you're referring to Git's handling of root commits or JJ's handling of root commits. Consider rephrasing to "JJ's root commits are the first ...".

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Thanks, I added that, and also added a note that it's quite different from Git's HEAD.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: > Suggestion: maybe visible is something like "can be returned by a revset"? I guess that's not entirely true, since you can return non-visible commits by querying by the commit id.

Yes, that's not quite true.

> Maybe visible is closer to "the set of commits reachable from the current view". Then to make it non-circular, we'd drop the word "visible" from the view section

There can actually be branches pointing to hidden commits. It's common to have a "remote-tracking branch" pointing to a hidden commits. If we didn't, we'd end up showing divergent changes quite often, or we'd have to define "visible" and "current" as two different sets or something. Btw, I don't want to rule out doing that, and I hope everyone feels free to question past decisions.

> Then to make it non-circular, we'd drop the word "visible" from the view section

Ooh, I like that idea. Done.

File: docs/glossary.md
Comment (listx) [reply docs/glossary.md]: The simplest thing that comes to mind is to make "revsets" refer to "revision sets" and "revset expressions" as the expressions of the functional language. If "revset expression" is too long, perhaps "revset query" (and you could name the functional language Revset Query Language).

File: docs/glossary.md
Comment (listx) [reply docs/glossary.md]: Love the examples. Thank you!

File: docs/glossary.md
Comment (listx) [reply docs/glossary.md]: LGTM

File: docs/glossary.md
Comment (ilyagr) [reply docs/glossary.md:22]: Seems fine.

File: docs/glossary.md
Comment (ilyagr) [reply docs/glossary.md]: I think you missed the last paragraph of my suggestion, where I reworded your description of hexadecimals. I still like the reworded version better, but feel free to go with either one.

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:66]: Should we mention that they coincide with Git's commit id when using the Git backend? Or do we not want to promise that?

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:73]: I'd add: "Most tools designed for Git can be easily used on such repositories. `jj` and `git` commands can be used interchangeably."

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:79]: This is nice! (No action needed)

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:116]: The last two clauses (after "much in the same way...") seem a bit confusing to me, especially the one about "commit history". I'll add something here if/once I figure out how I'd rewrite it.

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:122]: I'm again unsure how to improve this (will think about it), but I think it would be nice to more clearly explain the distinction that what an `hg` or `git` user would call a "repository" corresponds to "operation" here, while what we called a repository includes the entire history.

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:142]: I'd mention that "Some examples of rewriting a commit would be changing its description or rebasing it. Modifying the working copy rewrites the working copy commit."

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:164]: a "visible anonymous head"? In the definition of "Heads", you seemed to allow for these to not be visible.

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:165]: Perhaps add that "A working copy commit is always visible".

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:169]: I would put (visible), with or without parantheses, before anonymous heads here. This is related to a previous comment.

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:8]: Perhaps add something along the lines of:

Visible anonymous branches are tracked by the [view], which stores a list of [heads] of such branches.

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:180]: Maybe "a pointer to the associated repository"?

File: docs/glossary.md
Comment (ilyagr) [reply docs/glossary.md:116]: Here's an idea, to put after "The operation log is the
[DAG](https://en.wikipedia.org/wiki/Directed_acyclic_graph) formed by
[operation](#operation) objects."

When operations happen in sequence, they form a single line in the graph. Operations that happen concurrently from jj's perspective result in splits and/or merges in the DAG.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Done.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:66]: I think it makes sense to mention. If we ever change it, we'll just have to remember to update this text.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Good idea. Done.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:81]: Thanks :)

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:121]: Thanks, I added that with some minor tweaks.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:129]: I think of all operations as part of the repository. If someone says that their repository is corrupt, I would agree with them even if only some old operation is corrupt. I'd probably also agree even if just the working copy state is corrupt. These kinds of corruption of course have different levels of seriousness, but as I said, I think I'd consider them all "repo corruption".

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:149]: Good idea. Examples are helpful.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:173]: I was deliberately a bit vague in the definition of "heads". I've now made it more specific. Since all anonymous heads are visible (by definition), I don't think we need to add "visible" here.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:174]: I'm not sure that's strictly true, so I'd rather not say it.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:178]: It's actually only the "anonymous heads" that define visibility. They're actually called just "heads" in the model (https://github.com/martinvonz/jj/blob/cf402af9c15755064cc3cc435b2c02223e51354d/lib/src/op_store.rs#L193). "Visible heads" is technically more correct than "anonymous heads", but I found it hard to explain it without ending up with circular definitions. See some earlier discussion between me and @chooglen in this PR. Perhaps we should change back to "visible heads" if we can think of a better way to explain it.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:189]: I'm a little worried that would make the reader think that the repository always lives somewhere else. Isn't the second sentence in the paragraph enough to explain that they can be shared?

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Done

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: True, we could simply call the expressions "revset expression". I started making that change but I felt that it didn't help, mostly because we almost always mean "revset expression" when we say "revset", so it seems like it would be more useful to find a longer term for the set of revisions selected by a revset expression. Feel free to attempt it and send a PR to update the glossary and revsets.md once this PR has landed.

File: docs/glossary.md
Comment (ilyagr) [inline docs/glossary.md:42]: Very minor and optional: on second thought, perhaps it makes sense to write the range as "z-k" here to hint at correspondence.

File: docs/glossary.md
Comment (ilyagr) [reply docs/glossary.md:177]: Sounds good to postpone fixing this, once we figure out a better way. I think what you have now is good enough for people who care about such details to figure it out after a bit of confusion, and most people will only care about the gist of it anyway.

File: docs/glossary.md
Comment (ilyagr) [reply docs/glossary.md:128]: I think this terminology is very reasonable, it is just a little surprising.

I'm still thinking about a better way to phrase it, but it's certainly not a blocker.

File: docs/glossary.md
Comment (ilyagr) [reply docs/glossary.md:188]: I find this mildly confusing (if a workspace contains a repository, how could there be multiple workspaces per repository), but this is a minor issue. If we can't think of a clearly better phrasing, let's leave it as is for now.

**Review by ilyagr — APPROVED** [review]
I'll officially approve this, since I think it's more than good enough to be helpful to users. It might still make sense to wait for more people to comment or think of improvements, or you can submit it now and we can think about improvements in future PRs -- up to you.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md]: Done (here, but I left it as "k-z" on the line above since order seems unimportant there)

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:188]: I added a bit more detail here. PTAL. Btw, this is also related to the definition of "repository". In the source code and in the structure under `.jj/`, the repo (`.jj/repo/`) is separate from the working copy and state (`.jj/working_copy/`).

Comment (ilyagr) [reply docs/glossary.md:188]: Looks good, I think. Thanks!

> In the source code and in the structure under .jj/, the repo (.jj/repo/) is separate from the working copy and state (.jj/working_copy/).

Yes, I think that's a good distinction. The definition of a repository should really be "everything under `.jj/repo`" (I'm not proposing we change it right now in this doc, though -- if we do that and nothing else, it will probably only look more confusing). I think and hope people won't be confused by the explanation as it is now, but we can try to make it more precise later if we don't figure out a perfectly precise wording right now.

File: docs/glossary.md
Comment (chooglen) [inline docs/glossary.md:107]: (No action required)
This section is really well-worded :)

**Review by chooglen — APPROVED** [review]
LGTM. I'm not sure if the `#visible-commits` description is inaccurate or not, but even if it's inaccurate, I don't think it's egregious enough to block merging. It looks easy enough to fix post-review or in a follow-up PR.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:172]: > I thought `all()` also includes commits reachable by local branch targets but not remote branch targets. If so, then is `They are the commits that are reachable from an anonymous head in the [view](#view).` accurate?

No, `all()` are the commits reachable from the view's heads (https://github.com/martinvonz/jj/blob/c7b66606a181d77d1b18aad3e2f1fc6a0003451d/lib/src/op_store.rs#L193). Note that those heads don't necessarily include all local branch targets, although they practically always do. Also note that the set of visible heads is not produced by unioning the heads with branches and tags etc; the view's heads are exactly the visible heads.

Comment (ilyagr) [reply docs/glossary.md:172]: It turns out `jj branch s branch abandon_commit` works and doesn't make the commit visible. So, indeed, a branch that's not in `all()` can be created.

File: docs/glossary.md
Comment (elasticdog) [inline docs/glossary.md:40]:
```suggestion
16 bytes long and are often randomly generated. By default, `jj log` presents
```

Comment (martinvonz) [reply docs/glossary.md]: Fixed, thanks.

File: docs/glossary.md
Comment (martinvonz) [reply docs/glossary.md:172]: Good example :) The negative terms in branch conflicts are also not generally visible. Again, I'm not sure it was the right decision to do it this way, but it at least seems better than it was when both positive and negative terms were always added to the heads (which was 2 years ago, perhaps).

**Comment by martinvonz** [comment]
I sounds like this is good enough now, so I'll enable auto-merge. Any further changes can go into follow-up PRs.

File: docs/glossary.md
Comment (joyously) [reply docs/glossary.md:107]: I disagree that this is well-worded, because it mixes several things: parentage (by mentioning descendants), revsets, view, visibility, and the graph(or tree by mentioning branches).
It sounds to me like the DAG is referred to by tree terms, so why call it a `head` when it's a `leaf` in the tree? Only because Git does? (even though you say it's quite different from Git)

Comment (ilyagr) [reply docs/glossary.md:107]: I think you're both right :). This section is clear enough to show problems with the terminology we're using.

I also find the "anonymous head" terminology confusing, since 1) it's actually about visibility of commits and 2) if I understand correctly, it doesn't even have to be anonymous.

On that note, if I don't forget, I'll send in a correction to what seems to me to be a minor mistake here. Where it says "heads without a branch pointing to them", I think it should say "heads that may or may not have a named branch pointing to them".

I think it would be nice to come up with a better term. One possibility that comes to my mind is "visible head". "visible leaf" also would seem to work, but then we'd need to replace "head" with "leaf" throughout the code.

Comment (joyously) [reply docs/glossary.md:107]: I say there would be less confusion if you don't use the same term as Git, with a different meaning.

**Pull request merged**
